node/src/exchange/server.rs

```diff
@@ -1,52 +1,382 @@
     pub(super) fn on_response_sent(&mut self, peer: PeerId, request_id: RequestId) {
-        // TODO
+        trace!("response_sent; request_id: {request_id}, peer: {peer}");
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_failure(
         &mut self,
         peer: PeerId,
         request_id: RequestId,
         error: InboundFailure,
     ) {
-        // TODO
+        // TODO: cancel job if libp2p already failed it?
+        trace!("on_failure; request_id: {request_id}, peer: {peer}, error: {error:?}");
+    }
+
+    pub fn poll(&mut self, cx: &mut Context<'_>, sender: &mut R) -> Poll<()>
+    where
+        <R as ResponseSender>::Channel: std::fmt::Debug,
+    {
+        let pol = self.rx.poll_recv(cx);
+        if let Poll::Ready(Some((channel, response))) = pol {
+            sender.send_response(channel, response);
+            return Poll::Ready(());
+        }
+
+        Poll::Pending
+    }
+
+    fn handle_request_current_head(&mut self, channel: R::Channel) {
+        let store = self.store.clone();
+        let tx = self.tx.clone();
+
+        spawn(async move {
+            let response = store
+                .get_head()
+                .await
+                .map(|head| head.to_header_response())
+                .unwrap_or_else(|_| HeaderResponse::not_found());
+
+            let _ = tx.send((channel, vec![response])).await;
+        });
+    }
+
+    fn handle_request_by_hash(&mut self, channel: R::Channel, hash: Vec<u8>) {
+        let store = self.store.clone();
+        let tx = self.tx.clone();
+        spawn(async move {
+            let response = if let Ok(hash) = Hash::from_bytes(Algorithm::Sha256, &hash) {
+                store
+                    .get_by_hash(&hash)
+                    .await
+                    .map(|head| head.to_header_response())
+                    .unwrap_or_else(|_| HeaderResponse::not_found())
+            } else {
+                HeaderResponse::invalid()
+            };
+
+            let _ = tx.send((channel, vec![response])).await;
+        });
+    }
+
+    fn handle_request_by_height(&mut self, channel: R::Channel, origin: u64, amount: u64) {
+        let store = self.store.clone();
+        let tx = self.tx.clone();
+
+        spawn(async move {
+            let mut responses = vec![];
+            for i in origin..origin + amount {
+                let response = store
+                    .get_by_height(i)
+                    .await
+                    .map(|head| head.to_header_response())
+                    .unwrap_or_else(|_| HeaderResponse::not_found());
+                responses.push(response);
+            }
+
+            let _ = tx.send((channel, responses)).await;
+        });
+    }
+
+    fn handle_invalid_request(&self, channel: R::Channel) {
+        let tx = self.tx.clone();
+
+        spawn(async move {
+            let _ = tx.send((channel, vec![HeaderResponse::invalid()])).await;
+        })
+    }
+}
```

Comment:
Small optimization: We can `try_send` first and if it fails, then we do it thought a new task.

```rust
    fn handle_invalid_request(&self, channel: R::Channel) {
        if let Err(TrySendError::Full(val)) =
            self.tx.try_send((channel, vec![HeaderResponse::invalid()]))
        {
            let tx = self.tx.clone();

            spawn(async move {
                let _ = tx.send(val).await;
            });
        }
    }
```

Reply:
nice, added the `try_send`